proto/regen/ecocredit/v1beta1/marketplace/state.proto

```diff
@@ -0,0 +1,98 @@
+    // batch_denom is the credit batch being sold.
+    string batch_denom = 3;
+
+    // quantity is the quantity of credits being sold.
+    string quantity = 4;
+
+    // ask_price is the price the seller is asking for each unit of the
+    // batch_denom. Each credit unit of the batch will be sold for at least the
+    // ask_price or more.
+    cosmos.base.v1beta1.Coin ask_price = 5;
+
+    // disable_auto_retire disables auto-retirement of credits which allows a
+    // buyer to disable auto-retirement in their buy order enabling them to
+    // resell the credits to another buyer.
+    bool disable_auto_retire = 6;
+}
+
+// BuyOrder represents the information for a buy order.
+message BuyOrder {
+
+    option (cosmos.orm.v1alpha1.table) = {
+        id: 2,
+        primary_key: {fields: "buy_order_id"}
+        index: {id: 1 fields: "buyer"}
+    };
+
+    // Selection defines a buy order selection.
+    message Selection {
+
+        // sum defines the type of selection.
+        oneof sum {
+            // sell_order_id is the sell order ID against which the buyer is trying to buy.
+            // When sell_order_id is set, this is known as a direct buy order because it
+            // is placed directly against a specific sell order.
+            uint64 sell_order_id = 1;
+
+            // TODO: once we have filters defined for baskets, we can enable filtered
+            // buy orders which will involve full double-sided batch order matching
+            //
+            // filter selects credits to buy based upon the specified filter criteria.
+            // Filter filter = 2;
+        }
+    }
+
+    // buy_order_id is the unique ID of buy order.
+    uint64 buy_order_id = 1;
+
+    // buyer is the address that created the buy order
+    string buyer = 2;
+
+    // selection is the buy order selection.
+    Selection selection = 3;
+
+    // quantity is the quantity of credits to buy. If the quantity of credits
+    // available is less than this amount the order will be partially filled
+    // unless disable_partial_fill is true.
+    string quantity = 4;
+
+    // bid price is the bid price for this buy order. A credit unit will be
+    // settled at a purchase price that is no more than the bid price. The
+    // buy order will fail if the buyer does not have enough funds available
+    // to complete the purchase.
+    cosmos.base.v1beta1.Coin bid_price = 5;
+
+    // disable_auto_retire allows auto-retirement to be disabled. If it is set to true
+    // the credits will not auto-retire and can be resold assuming that the
+    // corresponding sell order has auto-retirement disabled. If the sell order
+    // hasn't disabled auto-retirement and the buy order tries to disable it,
+    // that buy order will fail.
+    bool disable_auto_retire = 6;
+
+    // disable_partial_fill disables the default behavior of partially filling
+    // buy orders if the requested quantity is not available.
+    bool disable_partial_fill = 7;
+}
```

Comment:
We need to add `AllowAskDenom` here as well.

Reply:
https://github.com/regen-network/regen-ledger/pull/700/commits/22491e9d5d575fadaa875d201750bf95933ea5c9